src/lib/utils/SynapseClient.tsx

```diff
@@ -608,3 +638,245 @@ export const signOut = () => {
+          (addPartResponse: AddPartResponse) => {
+            if (addPartResponse.addPartState === 'ADD_SUCCESS') {
+              // done with this part!
+              multipartUploadStatus.clientSidePartsState![partNumber - 1] = true
+              checkUploadComplete(
+                multipartUploadStatus,
+                fileName,
+                sessionToken,
+                fileUploadResolve,
+                fileUploadReject,
+                endpoint)
+            } else {
+              // retry after a brief delay
+              delay(1000).then(() => {
+                processFilePart(
+                  partNumber,
+                  multipartUploadStatus,
+                  sessionToken,
+                  fileName,
+                  file,
+                  request,
+                  fileUploadResolve,
+                  fileUploadReject,
+                  endpoint)
+              })
+            }
+          })
+      })
+    })
+}
+export const checkUploadComplete = (
+  status: MultipartUploadStatus,
+  fileHandleName: string,
+  sessionToken: string | undefined,
+  fileUploadResolve: (fileUpload: FileUploadComplete) => void,
+  fileUploadReject: (reason: any) => void,
+  endpoint: string = DEFAULT_ENDPOINT) => {
+  // if all client-side parts are true (uploaded), then complete the upload and get the file handle!
+  if (status.clientSidePartsState!.every((v) => { return v })) {
+    const url = `/file/v1/file/multipart/${status.uploadId}/complete`
+    doPut(url, undefined, sessionToken, undefined, endpoint).then((newStatus: MultipartUploadStatus) => {
+      // success!
+      fileUploadResolve({ fileHandleId: newStatus.resultFileHandleId, fileName: fileHandleName })
+    }).catch((error) => {
+      fileUploadReject(error)
+    })
+  }
+}
+const uploadFilePart = async (presignedUrl: string, file: any, contentType: string) => {
+  // TODO: could try using axios to get upload progress, then update the client-side part state (change to numbers from 0-1)
+  // This would give progress for the single file (across all parts).
+  // The parent would still need to figure out progress (for the total file set).
+  await fetch(presignedUrl, {
+    method: 'PUT',
+    headers: {
+      'Content-Type': contentType
+    },
+    body: file
+  })
+}
+export const startMultipartUpload = (
+  sessionToken: string | undefined,
+  fileName: string,
+  file: Blob,
+  request: MultipartUploadRequest,
+  fileUploadResolve: (fileUpload: FileUploadComplete) => void,
+  fileUploadReject: (reason: any) => void,
+  endpoint: string = DEFAULT_ENDPOINT) => {
+  const url = 'file/v1/file/multipart'
+  doPost(url, request, sessionToken, undefined, endpoint).then(async (status: MultipartUploadStatus) => {
+    // started the upload
+    // keep track of the part state client-side
+    const clientSidePartsState: boolean[] = []
+    for (let i = 0; i < status.partsState.length; i = i + 1) {
```

Comment:
this could get rewritten as 
const clientSidePartsState: boolean[] = status.partsState.split('').map(bit => bit === '1')